### PR TITLE
PICARD-2162: Improve performance of coverart box when selecting multiple files

### PR DESCRIPTION
**Pull request by Gabrielcarvfer**
# Summary

* This is a…
    * [ ] Bug fix
    * [ ] Feature addition
    * [x] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [ ] Other
* **Describe this change in 1-2 sentences**:
Make selection updates faster by comparing less stuff.

# Problem

When the selected items change, the metadatabox triggers an update of the coverart box, which compares and selects unique images to display.

* JIRA ticket (_optional_): PICARD-2162

# Solution

Replace the comparison of sets of images by a dictionary update. If the dictionary grows, new images were found and the 'changed' flag is set. If size remains the same, nothing changes.

# Action

### CODE REVIEW

**Comment by Gabrielcarvfer** [comment]
With 18k files selected, the time spent on picard.util.imagelist._update_state fell from around 1.5 s (pretty long freeze) to 90 ms (16x faster, which still is noticeable).

**Comment by Gabrielcarvfer** [comment]
Original
pre_changes
Switch from sets to dict
post_changes
Moving the Track import to the caller and passing it by argument to the callee (~30x speedup :D)
post_changes_2

File: picard/util/imagelist.py [hunk 1]
```diff
@@ -104,42 +119,46 @@ def __init__(self):
...
 
-def _process_images(state, src_obj):
-    from picard.track import Track
-
+def _process_images(state, src_obj, Track):
     # Check new images
     if state.update_new_metadata:
-        if state.new_images != set(src_obj.metadata.images):
+        src_dict = src_obj.metadata.images.hash_dict()
+        og_len = len(state.new_images)
```
Comment (zas) [inline picard/util/imagelist.py:126]: s/og_len/orig_len/?

**Review by zas — CHANGES_REQUESTED** [review]

File: picard/util/imagelist.py (hunk 1 above)
Comment (Gabrielcarvfer) [reply picard/util/imagelist.py]: Changed for the `s`. I didn't want to use `orig_len` because there are `metadata` and `orig_metadata`.

Comment (zas) [reply picard/util/imagelist.py]: Actually `s` is even less explicit (and `s` is used much more often for strings than lengths, if we were to keep one-char variable name, `l` would be more logicial).
What's `len(state.new_images)` ?

File: picard/util/imagelist.py [hunk 2]
```diff
@@ -104,42 +119,46 @@ def __init__(self):
...
 
-def _process_images(state, src_obj):
-    from picard.track import Track
-
+def _process_images(state, src_obj, Track):
     # Check new images
     if state.update_new_metadata:
-        if state.new_images != set(src_obj.metadata.images):
+        src_dict = src_obj.metadata.images.hash_dict()
+        s = len(state.new_images)
```
Comment (zas) [inline picard/util/imagelist.py:126]: What about `pre_update_length` or `prev_len`?

**Review by zas — CHANGES_REQUESTED** [review]

File: picard/util/imagelist.py (hunk 2 above)
Comment (Gabrielcarvfer) [reply picard/util/imagelist.py:126]: Way better.

**Review by zas — APPROVED** [review]

**Review by phw — APPROVED** [review]

**Pull request merged**